### ISSUE TITLE
Issue when invoking asynchronous call stack .NET/C#

### ISSUE THREAD

**Issue by upsilondynamics** [Microsoft/ClearScript#104]
I am having an issue executing an asynchronous method from the script, if I have an async call from the script and another asynchronous call from within that method, it doesn't await.  But If I have no async call inside my C# method, it waits in the JS code.  Any help would be appreciated.

Javascript side:

//JS Script to call async function
```
function callAsync(t) { return await t; }
var result = callAsync(myAsyncFunction(()); //Finishes BEFORE SomeOtherFunction call does below
```

C# Code (WONT AWAIT -> The .Execute call on the v8engine object finishes)
```
public async Task<string> myAsyncFunction() {
   var result = await SomeOtherFunction();
   return result; //This point in the code will be hit after the Execute finishes on the v8scriptengine object
}
```

C# Code (WILL await)
```
public async Task<string> myAsyncFunction() {
   return "Some Result";
}
```

In other words, single level asynchronous call works - calling a method which in turn await another method, etc and down the entire call stack just causes the await to continue on the clearscript side without waiting.  

I've spent way too many hours trying to get this to work, any help please :)

**Comment by ClearScriptLib** [comment]
Hi @upsilondynamics,

JavaScript's [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await) doesn't work with tasks as you're expecting. When given something that isn't a [`promise`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise), it simply converts it into a _resolved_ promise, in this case one whose value is the task.

Clearly that's not very useful. One way to make it work is to provide a method that wraps a promise around a task. Consider something like this (requires [ClearScript 5.5.5](https://www.nuget.org/packages/Microsoft.ClearScript/5.5.5)):

``` C#
public static class TaskScripting {
    private delegate void Executor(dynamic resolve, dynamic reject);
    public static object ToPromise<T>(this Task<T> task) {
        var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
        return ctor.Invoke(true, new Executor((resolve, reject) => {
            task.ContinueWith(t => {
                if (t.IsCompleted) {
                    resolve(t.Result);
                }
                else {
                    reject(t.Exception);
                }
            });
        }));
    }
}
```

You could add this class to your engine:

``` C#
engine.AddHostType(typeof(TaskScripting));
```

And then use the conversion method in your script code:

``` JavaScript
// JavaScript
callAsync(myAsyncFunction().ToPromise());
```

Good luck!

**Comment by upsilondynamics** [comment]
Thanks for the response, unfortunately it still isn't working.  The script ends before the await completes.

The **ToPromise** extension method does return the promise object, but even when I await the call when passing it into an async function - it still isn't working:

Javascript
`function async callAsync(t) { return await t; }`
callAsync(myMethod().ToPromise()); //This will just blow over way before the C# async finishes even with the extension
//POINT 1 REACHED

**C#**
```
 public async Task<string> myMethod()
{
 return await SomeOtherMethod();
//POINT 2 REACHED AFTER
}
```

**Comment by ClearScriptLib** [comment]
Hi @upsilondynamics,

What you're describing appears to be the correct behavior.

You're invoking `callAsync` from the (synchronous) root scope, so the invocation ends immediately and returns a promise that represents `callAsync`'s asynchronous operation.

Can you clarify how this differs from what you're expecting?

Thanks!

**Comment by upsilondynamics** [comment]
Example code:

Let me define a promise, with an asynchronous function to call it with a 5000 MS delay, invoke it from t the root.
```
wait =  (time) => new Promise(resolve => setTimeout(resolve, time));
async function testWait(delay){ await wait(delay);}
Console.WriteLine("Line 1");
testWait(5000); 
Console.WriteLine("Line 2");
```
There is zero delay between line 1 and line 2, I guess this makes sense becayse we are calling an asynchronous method from a synchronous root.  But even if I wrap it like such:

`(async () => { await testWait(5000); })();`

It still doesn't work, how am I supposed to get the result from the .NET side asynchronous call if it just blasts pasts the ScriptEngine execute call, that by the time the asynchronous call finishes, the script has long already been parsed and executed by Clearscript when I need those resultant values in the script itself?  

Meanwhile in C# Land:
_engine.Execute(theScript); //theScript == the JS code shown above *breakpoint here, HIT F5
System.Diagnostics.Debug.WriteLine("Finished"); // WE GO  HERE IMMEDIATELY, NOT AFTER 5 SECONDS

Why is this? If we run the same code in a browser, it will delay 5 seconds. Maybe I'm not understanding, but is there a way to await from the root?

**Comment by upsilondynamics** [comment]
**UPDATE**

I am going to change my methods to non-asynchronous since this is server side, the results must be ready before the request returns to the browser. That being said I think you guys should investigate implementing a proper **await** implementation so the main thread waits when an await is hit, and the result is returned before continuing onto the next script statement - that's how await/async should work with newer JS implementations.

**Comment by ClearScriptLib** [comment]
Hi @upsilondynamics,

Again, what you're describing is correct behavior.

JavaScript's `await` operator pauses asynchronous execution only. What it really does under the covers is add a hidden continuation callback to its operand (unless its operand isn't a promise, in which case it converts it into a resolved promise first). That callback _resumes_ the `async` function at a point just beyond the `await` and continues execution until the next `await` or `return`.

Simply invoking a JavaScript `async` function (as opposed to `await`-ing it) kicks off the asynchronous operation and returns immediately. This is by design. ClearScript isn't responsible for this behavior, nor can it change it.

Yes, you can [`Wait`](https://docs.microsoft.com/en-us/dotnet/api/system.threading.tasks.task.wait) for a .NET task synchronously, but you can't do the equivalent with a JavaScript promise. That's because a task can be completed on another thread, whereas JavaScript code is single-threaded.

One thing you may not realize is that a JavaScript `async` function can be resumed long after `ScriptEngine.Execute` has returned. Here's a command-line sample that demonstrates the sort of "multi-level" asynchronous operation we've been discussing:

``` C#
using (var engine = new V8ScriptEngine()) {
    engine.AddHostType(typeof(Console));
    engine.AddHostType(typeof(Task));
    engine.AddHostType(typeof(TaskScripting));
    engine.Script.done = new ManualResetEventSlim();
    engine.Execute(@"
        async function waitAsync(msec) {
            await Task.Delay(msec).ToPromise();
        }
        async function waitAndLog(msec, text) {
            await waitAsync(msec);
            Console.WriteLine(text);
        }
        async function runTestAsync() {
            await waitAndLog(1000, 'first');
            await waitAndLog(1000, 'second');
            await waitAndLog(1000, 'third');
            await waitAsync(1000);
            done.Set();
        }
        runTestAsync();
    ");
    engine.Script.done.Wait();
}
```

Note that, to make this work, we had to add a non-generic `ToPromise` overload:

``` C#
public static object ToPromise(this Task task) {
    var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
    return ctor.Invoke(true, new Executor((resolve, reject) => {
        task.ContinueWith(t => {
            if (t.IsCompleted) {
                resolve();
            }
            else {
                reject(t.Exception);
            }
        });
    }));
}
```

Cheers!

**Comment by upsilondynamics** [comment]
This was the magic bullet I was looking for (The ManualResetEventSlim class)
`engine.Script.done.Wait();`

This will halt execution on the .NET side until the promise is resolved, THANK YOU! Always can count on the geniuses at Microsoft.

Cheers.

**Comment by themightylc** [comment]
Hi @ClearScriptLib 

do you have a quick solution for this, when my Task (i.e. Promise) has to return a value? So Task<> instead of only Task?

> Note that, to make this work, we had to add a non-generic `ToPromise` overload:
> 
> ```cs
> public static object ToPromise(this Task task) {
>     var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
>     return ctor.Invoke(true, new Executor((resolve, reject) => {
>         task.ContinueWith(t => {
>             if (t.IsCompleted) {
>                 resolve();
>             }
>             else {
>                 reject(t.Exception);
>             }
>         });
>     }));
> }

**Comment by themightylc** [comment]
got this resolved by adding another ToStringPromise Extension that passes the Task result to the resolve delegate. 
Overloading the ToPromise Extension did not work, neither did using Task<Object>.
This works for me as I will mostly be passing JSON anyway. :)

> 
> > Note that, to make this work, we had to add a non-generic `ToPromise` overload:
> > ```cs
> > public static object ToStringPromise(this Task<string> task) {
> >     var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
> >     return ctor.Invoke(true, new Executor((resolve, reject) => {
> >         task.ContinueWith(t => {
> >             if (t.IsCompleted) {
> >                 resolve(t.Result);
> >             }
> >             else {
> >                 reject(t.Exception);
> >             }
> >         });
> >     }));
> > }
> > ```

**Comment by ClearScriptLib** [comment]
Hi @themightylc,

The following should work for all tasks:

``` C#
public static class TaskScripting {
    private delegate void Executor(dynamic resolve, dynamic reject);
    public static object ToPromise<T>(this Task<T> task) {
        var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
        return ctor.Invoke(true, new Executor((resolve, reject) => {
            task.ContinueWith(t => {
                if (t.IsCompleted) {
                    resolve(t.Result);
                }
                else {
                    reject(t.Exception);
                }
            });
        }));
    }
    public static object ToPromise(this Task task) {
        var ctor = (ScriptObject)ScriptEngine.Current.Script.Promise;
        return ctor.Invoke(true, new Executor((resolve, reject) => {
            task.ContinueWith(t => {
                if (t.IsCompleted) {
                    resolve();
                }
                else {
                    reject(t.Exception);
                }
            });
        }));
    }
}
```

Here's a usage sample:

``` C#
engine.AddHostType(typeof(TaskScripting));
engine.AddHostType(typeof(Console));

engine.Script.voidTask = Task.Run(() => {});
engine.Script.stringTask = Task.Run(() => "Hello, world!");

engine.Execute(@"
    voidTask.ToPromise().then(() => Console.WriteLine('Void task completed.'));
    stringTask.ToPromise().then(msg => Console.WriteLine(msg));
");
```

Good luck!
